runtime/ts/storage/in-memory-storage.ts

```diff
@@ -155,22 +148,41 @@ abstract class InMemoryStorageProvider extends StorageProviderBase {
     }
     return new InMemoryVariable(type, storageEngine, name, id, key);
   }
+
+  // A consequence of awaiting this function is that this.backingStore
+  // is guaranteed to exist once the await completes. This is because
+  // if backingStore doesn't yet exist, the assignment in the then()
+  // is guaranteed to execute before anything awaiting this function.
+  async ensureBackingStore() {
+    if (this.backingStore) {
+      return this.backingStore;
+    }
+    if (!this.pendingBackingStore) {
+      const key = this.storageEngine.baseStorageKey(this.backingType());
+      this.pendingBackingStore = this.storageEngine.baseStorageFor(this.type, key);
+      this.pendingBackingStore.then(backingStore => this.backingStore = backingStore);
+    }
+    return this.pendingBackingStore;
```

Comment:
Nice rework. One possible concern - does this mean the pendingBackingStorePromise will always hang around after the backing store has been resolved? Not a huge problem, but might be good to discard the memory for it.